core/src/main/java/org/geysermc/geyser/session/cache/tags/GeyserHolderSet.java

```diff
@@ -0,0 +1,116 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.geysermc.geyser.session.GeyserSession;
+import org.geysermc.geyser.session.cache.TagCache;
+import org.geysermc.geyser.session.cache.registry.JavaRegistryKey;
+
+/**
+ * Similar to vanilla Minecraft's HolderSets, stores either a tag or a list of IDs (this list can also be represented as a single ID in vanilla HolderSets).
+ *
+ * <p>Because HolderSets utilise tags, when loading a HolderSet, Geyser must store tags for the registry the HolderSet is for (see {@link JavaRegistryKey}).</p>
+ *
+ * <p>Use the {@link GeyserHolderSet#readHolderSet} method to easily read a HolderSet from NBT sent by a server. To turn the HolderSet into a list of network IDs, use the {@link GeyserHolderSet#resolveRaw} method.
+ * To turn the HolderSet into a list of objects, use the {@link GeyserHolderSet#resolve} method.</p>
+ */
+@Data
+public final class GeyserHolderSet<T> {
+
+    private final JavaRegistryKey<T> registry;
+    private final @Nullable Tag<T> tag;
+    private final int @Nullable [] holders;
+
+    public GeyserHolderSet(JavaRegistryKey<T> registry, int @NonNull [] holders) {
+        this.registry = registry;
+        this.tag = null;
+        this.holders = holders;
+    }
+
+    public GeyserHolderSet(JavaRegistryKey<T> registry, @NonNull Tag<T> tagId) {
+        this.registry = registry;
+        this.tag = tagId;
+        this.holders = null;
+    }
+
+    /**
+     * Resolves the HolderSet, and automatically maps the network IDs to their respective object types. If the HolderSet is a list of IDs, this will be returned. If it is a tag, the tag will be resolved from the tag cache.
+     *
+     * @return the HolderSet turned into a list of objects.
+     */
+    public List<T> resolve(GeyserSession session) {
+        return TagCache.mapRawArray(session, resolveRaw(session.getTagCache()), registry);
+    }
+
+    /**
+     * Resolves the HolderSet. If the HolderSet is a list of IDs, this will be returned. If it is a tag, the tag will be resolved from the tag cache.
+     *
+     * @return the HolderSet turned into a list of objects.
+     */
+    public int[] resolveRaw(TagCache tagCache) {
+        if (holders != null) {
+            return holders;
+        }
+
+        return tagCache.getRaw(Objects.requireNonNull(tag, "HolderSet must have a tag if it doesn't have a list of IDs"));
+    }
+
+    /**
+     * Reads a HolderSet from an object from NBT.
+     *
+     * @param registry the registry the HolderSet contains IDs from.
+     * @param holderSet the HolderSet as an object from NBT.
+     * @param keyIdMapping a function that maps resource location IDs in the HolderSet's registry to their network IDs.
+     */
+    public static <T> GeyserHolderSet<T> readHolderSet(JavaRegistryKey<T> registry, @Nullable Object holderSet, Function<Key, Integer> keyIdMapping) {
+        if (holderSet == null) {
+            return new GeyserHolderSet<>(registry, new int[]{});
```

Comment:
looks like this could also be an IntArrays.EMPTY_ARRAY

Reply:
Yep, just noticed it too